business/dashboards.go

```diff
@@ -39,12 +39,28 @@ func NewDashboardsService() *DashboardsService {
 	if nsLabel == "" {
 		nsLabel = "kubernetes_namespace"
 	}
+
+	// Overwrite Custom dashboards defined at Namespace level
+	builtInDashboards := cfg.CustomDashboards.OrganizeByName()
+	if namespace != nil {
+		nsDashboards := dashboards.GetNamespaceMonitoringDashboards(namespace.Name, namespace.Annotations)
+		for name, dashboard := range nsDashboards.OrganizeByName() {
+			builtInDashboards[name] = dashboard
+		}
+	}
+	if workload != nil {
+		wkDashboards := dashboards.GetWorkloadMonitoringDashboards(namespace.Name, workload.Name, workload.DashboardAnnotations)
+		for name, dashboard := range wkDashboards.OrganizeByName() {
+			builtInDashboards[name] = dashboard
+		}
+	}
+
```

Comment:
OK. Almost there. The UI looks OK now. However, the [molecule test](https://github.com/kiali/kiali-operator/pull/352/files#diff-44588589631a74bd989b470bcf4a3c449be1939cd1d55c781098ed6ff5aa790dR115) is failing because custom_envoy1 still shows up (returns 200 status code), though it is empty. This API should return a 500 and say the dashboard does not exist but instead it returns this:

```
$ curl "http://localhost:20001/kiali/api/namespaces/simple-mesh/customdashboard/custom_envoy1?labelsFilter=app:simple-server,version:v99&workloadType=Deployment&workload=simple-server"

{"name":"custom_envoy1","title":"","charts":[],"aggregations":[],"externalLinks":[]}
```

The reason for this is the way the dashboards are aggregated in this PR is not correct. This code in the PR is not using function [AddMonitoringDashboards](https://github.com/kiali/kiali/blob/v1.36.0/config/dashboards/dashboards_config.go#L99-L103) but it should.

The way it should be working is - if a dashboard is declared but has NO charts and NO discoverOn then that dashboard should be removed [as you see commented here](https://github.com/kiali/kiali/blob/v1.36.0/config/dashboards/dashboards_config.go#L102) and [implemented here](https://github.com/kiali/kiali/blob/v1.36.0/config/dashboards/dashboards_config.go#L114). We use this feature already when users declare dashboards in the Kiali CR in order to override the built-in dashboards. A user can REMOVE a built-in dashboard by declaring the dashboard by name only (with no additional data). We test this kind of thing already... see our [current molecule test](https://github.com/kiali/kiali-operator/blob/v1.36.0/molecule/api-test/dashboards.test.yaml#L87-L98) that checks for that functionality of removing a built-in dashboard ([this is what the yaml looks like](https://github.com/kiali/kiali-operator/blob/v1.36.0/molecule/api-test/kiali-cr.yaml#L11-L12) when removing a dashboard). The same thing should happen up the chain - I should be able to remove built-in dashboards (and those declared by the Kiali CR) by overriding it in the namespace annotation. And individual workloads should be able to remove dashboards even declared in namespace annotations by overriding it in the workload annotations (and this is what my molecule test is testing and failing).

The solution is to make this change:

```suggestion
	// Overwrite Custom dashboards defined at Namespace level and Workoad level
	builtInDashboards := cfg.CustomDashboards
	if namespace != nil {
		nsDashboards := dashboards.GetNamespaceMonitoringDashboards(namespace.Name, namespace.Annotations)
		builtInDashboards = dashboards.AddMonitoringDashboards(builtInDashboards, nsDashboards)
	}
	if workload != nil {
		wkDashboards := dashboards.GetWorkloadMonitoringDashboards(namespace.Name, workload.Name, workload.DashboardAnnotations)
		builtInDashboards = dashboards.AddMonitoringDashboards(builtInDashboards, wkDashboards)
	}
```

This requires an additional change down a few lines -- on line 63, you need to change that to be the following (I'll add a suggested change comment next - github won't let me do a single suggested change here):

```
dashboards:      builtInDashboards.OrganizeByName(),
```

With this fixed, things work:

```
$ curl "http://localhost:20001/api/namespaces/simple-mesh/customdashboard/custom_envoy1?labelsFilter=app:simple-server,version:v99&workloadType=Deployment&workload=simple-server"

{"error":"Dashboard [custom_envoy1] does not exist or is disabled"}
```

Reply:
Mmm ok, I was using the premise that the builtin were re-built on every query, so, if the user remove them, then they are not present and they should fall in the default.

I've added your suggestions, let me know how it looks like.

I'll continue on Monday, also this one would require some good QE testing as well.